### PR TITLE
Fix #1506 - Fix unsigned integer overflow in RzIO cache skyline

### LINKED ISSUE #1506: io.cache: oob-read when reading from cache
See https://github.com/radareorg/radare2/issues/19006 and https://github.com/radareorg/radare2/pull/19008/

### PR DESCRIPTION
**Pull request by GustavoLCR**
* Add test

 

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

`RzSkyline` needs a `RzInterval` without wrap around.

**Test plan**

Test pass

**Closing issues**

Fixes #1506

### CODE REVIEW

**Review by ret2libc — COMMENTED** [review]
Would it make sense to have this overflow condition also in rz_skyline API (with rz_return_val_if_fail) so that if someone calls it with the wrong ranges we get a warning at least?

**Review by ret2libc — CHANGES_REQUESTED** [review]
Code looks good and I confirm it fixes the problem, thanks! I'd just wait for the rz_return_ on rz_skyline_add and rz_skyline_get_item_intersect.

**Comment by XVilka** [comment]
@ret2libc good idea on overflow check in the API. If this possible it indeed might be helpful.

**Review by wargio — APPROVED** [review]

**Comment by XVilka** [comment]
```
test_rz_io_mapsplit ERR
[XX] Fail at line 64: UT64_MAX not mapped: expected true, got false

test_rz_io_mapsplit2 ERR
[XX] Fail at line 81: UT64_MAX not mapped: expected true, got false

test_rz_io_mapsplit3 ERR
[XX] Fail at line 94: UT64_MAX - 1 not mapped: expected true, got false


--- stderr ---
WARNING: rz_skyline_add: assertion 'skyline && !UT64_ADD_OVFCHK(rz_itv_begin(itv), rz_itv_size(itv))' failed (line 13)
WARNING: rz_skyline_add: assertion 'skyline && !UT64_ADD_OVFCHK(rz_itv_begin(itv), rz_itv_size(itv))' failed (line 13)
WARNING: rz_skyline_add: assertion 'skyline && !UT64_ADD_OVFCHK(rz_itv_begin(itv), rz_itv_size(itv))' failed (line 13)
-------
```

**Comment by XVilka** [comment]
```
[XX] db/cmd/cmd_px io.unalloc, io.unalloc.ch and pxc
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '. scripts/palette.rz
e io.unalloc=true
(dump; xc 0x00010060-0x0000ffd0 @ 0x0000ffd0-6)
e scr.color=3
e hex.comments=true
.(dump)
?e
e scr.color=0
.(dump)
?e
e hex.comments=false
e io.unalloc.ch=?
.(dump)
?e
e io.unalloc.ch=
.(dump)
' bins/elf/analysis/tiny.elf
-- stdout
--- expected
+++ actual
@@ -2,7 +2,7 @@
 0x0000ffca  .... .... .... .... .... .... .... ....                  
 0x0000ffda  .... .... .... .... .... .... .... ....                  
 0x0000ffea  .... .... .... .... .... .... .... ....                  
-0x0000fffa  .... .... .... 7f45 4c46 0100 0000 0000        .ELF......  ; segment.ehdr  ; [01] -rw- segment size 45 named ehdr
+0x0000fffa  .... .... ..ff 7f45 4c46 0100 0000 0000       ..ELF......  ; segment.ehdr  ; [01] -rw- segment size 45 named ehdr
 0x0001000a  0000 0000 0100 0200 0300 2000 0100 2000  .......... ... .
 0x0001001a  0100 0400 0000 b32a 31c0 40cd 8000 3400  .......*1.@...4.  ; entry0
 0x0001002a  2000 0100 0000 0000 0000 0000 0000 0000   ...............
@@ -13,7 +13,7 @@
 0x0000ffca  .... .... .... .... .... .... .... ....                  
 0x0000ffda  .... .... .... .... .... .... .... ....                  
 0x0000ffea  .... .... .... .... .... .... .... ....                  
-0x0000fffa  .... .... .... 7f45 4c46 0100 0000 0000        .ELF......  ; segment.ehdr  ; [01] -rw- segment size 45 named ehdr
+0x0000fffa  .... .... ..ff 7f45 4c46 0100 0000 0000       ..ELF......  ; segment.ehdr  ; [01] -rw- segment size 45 named ehdr
 0x0001000a  0000 0000 0100 0200 0300 2000 0100 2000  .......... ... .
 0x0001001a  0100 0400 0000 b32a 31c0 40cd 8000 3400  .......*1.@...4.  ; entry0
 0x0001002a  2000 0100 0000 0000 0000 0000 0000 0000   ...............
@@ -24,7 +24,7 @@
 0x0000ffca  ???? ???? ???? ???? ???? ???? ???? ????                  
 0x0000ffda  ???? ???? ???? ???? ???? ???? ???? ????                  
 0x0000ffea  ???? ???? ???? ???? ???? ???? ???? ????                  
-0x0000fffa  ???? ???? ???? 7f45 4c46 0100 0000 0000        .ELF......
+0x0000fffa  ???? ???? ??ff 7f45 4c46 0100 0000 0000       ..ELF......
 0x0001000a  0000 0000 0100 0200 0300 2000 0100 2000  .......... ... .
 0x0001001a  0100 0400 0000 b32a 31c0 40cd 8000 3400  .......*1.@...4.
 0x0001002a  2000 0100 0000 0000 0000 0000 0000 0000   ...............
@@ -35,7 +35,7 @@
 0x0000ffca                                                           
 0x0000ffda                                                           
 0x0000ffea                                                           
-0x0000fffa                 7f45 4c46 0100 0000 0000        .ELF......
+0x0000fffa              ff 7f45 4c46 0100 0000 0000       ..ELF......
 0x0001000a  0000 0000 0100 0200 0300 2000 0100 2000  .......... ... .
 0x0001001a  0100 0400 0000 b32a 31c0 40cd 8000 3400  .......*1.@...4.
 0x0001002a  2000 0100 0000 0000 0000 0000 0000 0000   ...............
[XX] db/cmd/cmd_print io.unalloc, io.unalloc.ch and prc
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '. scripts/palette.rz
(print; prc 0x00010032-0x0000fffa @ 0x0000fffa)
e io.unalloc=true
e scr.color=0
.(print)
?e
e io.unalloc.ch=-
.(print)
?e
e scr.color=3
e io.unalloc.ch=
.(print)
?e
e io.unalloc.ch=?
.(print)
' bins/elf/analysis/tiny.elf
-- stdout
--- expected
+++ actual
@@ -1,19 +1,19 @@
-0x0000fffa ????????????se,,,,,,            
+0x0000fffa ??????????##se,,,,,,            
 0x0001000a                     ..      ..  
 0x0001001a             en....OO,,OO;;  ..  
 0x0001002a ..              
 
-0x0000fffa ------------se,,,,,,            
+0x0000fffa ----------##se,,,,,,            
 0x0001000a                     ..      ..  
 0x0001001a             en....OO,,OO;;  ..  
 0x0001002a ..              
 
-0x0000fffa ..........[0m..se                  
+0x0000fffa ..........[38;2;0;0;0m  se                  
 0x0001000a                                 
 0x0001001a             en                  
 0x0001002a                 
 
-0x0000fffa ??????????[0m??se                  
+0x0000fffa ??????????[38;2;0;0;0m  se                  
 0x0001000a                                 
 0x0001001a             en                  
 0x0001002a                 

-- stderr
WARNING: Failed to read ELF header (e_phnum).
WARNING: Failed to read ELF header (e_shentsize).
WARNING: Failed to read ELF header (e_shnum).
WARNING: Failed to read ELF header (e_shstrndx).
WARNING: The binary seems to be a tiny elf (45 bytes). Reload e_phnum value.
WARNING: The segment 0 at 0x4 seems to be invalid.
WARNING: Failed to initialize section header.
```

File: librz/util/skyline.c [hunk 1]
```diff
@@ -10,7 +10,7 @@
 	(((addr) < (rz_itv_end(((RzSkylineItem *)(part))->itv)) || !rz_itv_end(((RzSkylineItem *)(part))->itv)) ? -1 : 1)
 
 RZ_API bool rz_skyline_add(RzSkyline *skyline, RzInterval itv, void *user) {
-	rz_return_val_if_fail(skyline, false);
+	rz_return_val_if_fail(skyline && (!rz_itv_size(itv) || !UT64_ADD_OVFCHK(rz_itv_begin(itv), rz_itv_size(itv) - 1)), false);
```
Comment (XVilka) [inline librz/util/skyline.c]: I think it's probably to extract this long check into multiple `rz_return_val_if_fail()` for readability.

File: librz/util/skyline.c [hunk 2]
```diff
@@ -54,13 +54,14 @@ RZ_API bool rz_skyline_add(RzSkyline *skyline, RzInterval itv, void *user) {
 }
 
 RZ_API const RzSkylineItem *rz_skyline_get_item_intersect(RzSkyline *skyline, ut64 addr, ut64 len) {
-	rz_return_val_if_fail(skyline, NULL);
+	rz_return_val_if_fail(skyline && len && !UT64_ADD_OVFCHK(addr, len - 1), NULL);
```
Comment (XVilka) [inline librz/util/skyline.c]: Probably better to split this `rz_return_val_if_fail()` in two.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/util/skyline.c [hunk 3]
```diff
@@ -11,6 +11,10 @@
 
 RZ_API bool rz_skyline_add(RzSkyline *skyline, RzInterval itv, void *user) {
 	rz_return_val_if_fail(skyline, false);
+	if (rz_itv_size(itv)) {
+		// Allow size 1 itv at UT64_MAX, block all other overflows
```
Comment (ret2libc) [inline librz/util/skyline.c]: @GustavoLCR why this? Does this block itv.addr == UT64_MAX-1, len = 2 as well?

File: test/unit/test_io.c [hunk 4]
```diff
@@ -14,23 +14,25 @@ bool test_rz_io_cache(void) {
...
+	mu_assert_true(rz_io_cache_write(io, 4, (ut8 *)"CD", 2), "Overlapped cache write at 4 failed");
+	mu_assert_true(rz_io_cache_write(io, 8, (ut8 *)"EFG", 3), "Cache write at 8 failed");
 	mu_assert_true(rz_io_cache_read(io, 0, buf, 2), "Cache read at 0 failed");
 	mu_assert_true(rz_io_cache_read(io, 2, buf + 2, 2), "Cache read at 2 failed");
 	mu_assert_true(rz_io_cache_read(io, 4, buf + 4, 2), "Cache read at 4 failed");
-	mu_assert_true(rz_io_cache_read(io, 6, buf + 6, 2), "Cache read at 6 failed");
-	mu_assert_true(rz_io_cache_read(io, 8, buf + 8, 3), "Cache read at 8 failed");
+	mu_assert_false(rz_io_cache_read(io, 6, buf + 6, 2), "Cache read at 6 should fail");
+	mu_assert_true(rz_io_cache_read(io, 7, buf + 7, 2), "Cache read at 7 with size 2 should succeed");
+	mu_assert_true(rz_io_cache_read(io, 9, buf + 9, 2), "Cache read at 8 failed");
```
Comment (ret2libc) [inline test/unit/test_io.c]:
```suggestion
	mu_assert_true(rz_io_cache_read(io, 9, buf + 9, 2), "Cache read at 9 failed");
```

File: librz/util/skyline.c (hunk 3 above)
Comment (ret2libc) [reply librz/util/skyline.c]: Judging from the unit tests you wrote, it seems that UT64_MAX-1, len=2 is possible indeed. Maybe it's just the comment wrong?

**Review by ret2libc — APPROVED** [review]
Amazing!

**Pull request merged**